README.md
ZfrStripe
=========

[![Latest Stable Version](https://poser.pugx.org/zfr/zfr-stripe/v/stable.png)](https://packagist.org/packages/zfr/zfr-stripe)

ZfrStripe is a modern PHP library based on Guzzle for [Stripe payment system](https://www.stripe.com).

## Dependencies

* PHP 5.4+
* [Guzzle](http://www.guzzlephp.org): >= 3.6

## Installation

Installation of ZfrStripe is only officially supported using Composer:

```sh
php composer.phar require 'zfr/zfr-stripe:3.*'
```

## Tutorial

First, you need to instantiate the Stripe client, passing your private API key (you can find it in your Stripe
settings):

```php
$client = new StripeClient('my-api-key');
```

> You can change the API key for the client using the `setApiKey` method. This is useful if you are using Stripe
Connect and make both your own API calls and API calls on behalf of your users.

The currently latest supported version of the API is **2015-10-16**. You can (and should) also explicitly specify the version
of the client using the second parameter:

```php
$client = new StripeClient('my-api-key', '2015-10-16');
```

### Versioning

Stripe API has an odd (yet very useful) [versioning policy](https://stripe.com/docs/upgrades) that makes it hard to version it correctly. The trick
is that Stripe may change how some properties are passed on return on new versions (hence being a BC for older version), but still introduce new endpoints
that will be available to older versions too.

Therefore, each time a new dated version of Stripe API is released, ZfrStripe will adopt this policy:

* If new endpoints are introduced, without changing old ones, the new endpoints are added for all supported versions. A minor version is tagged, and set
the default Stripe version as the latest one.
* If no new endpoints are introduced, but some parameter change, a new descriptor is created, hence supporting multiple versions. A minor version is tagged, 
and set the default Stripe version as the latest one.
* If no new endpoints are introduced, and no parameter are changed, a minor version is tagged, and set the default Stripe version as the latest one.
* If existing endpoints are updating by changing their URL, a major release of ZfrStripe is released as compatibility cannot be assured.

Currently, the following Stripe API versions are accepted by ZfrStripe: `2015-02-18`, `2015-03-24`, `2015-04-07`, `2015-06-15`, `2015-07-07`, 
`2015-07-13`, `2015-07-28`, `2015-08-07`, `2015-08-19`, `2015-09-03`, `2015-09-08`, `2015-09-23`, `2015-10-01`, `2015-10-12`,
`2015-10-16`. I will try to update the library as soon as new version are released.

> If you need support for versions as old as 2014-03-28, please use branch v2 of ZfrStripe.
> If you need support for even older versions, please use branch v1 of ZfrStripe.

### How to use it?

Using the client is easy. For instance, here is how you'd create a new charge:

```php
$details = $client->createCharge([
    'amount'   => 500,
    'currency' => 'EUR',
    'customer' => 'cus_37EGGf4LMGgYy8'
]);
```

The parameters have a direct one-to-one mapping with the official documentation (for any reference, you can also
check the `ZfrStripe\Client\ServiceDescription\Stripe-v1.1.php` file). To know what the responses look like, please
refer to the [official API reference](https://stripe.com/docs/api).

#### Using a different Stripe API version

For each requests, we send the "Stripe-Version" header to Stripe. Usually, the best
way is to upgrade your API version globally in your Stripe dashboard. However, you may want to use two different versions in different part of your code, or test new features during development without changing it globally to your whole Stripe account.

To do that, you can explicitely set the wanted version in the constructor. Example:

```php
$stripeClient = new StripeClient('my-api-key', '2014-03-28');
```

You can dynamically change the version using the `setApiVersion` method:

```php
$stripeClient = new StripeClient('my-api-key', '2014-03-28');

// Responses will be formatted according to the 2014-03-28 version...

$stripeClient->setApiVersion('2014-08-20');

// Responses will now be formatted according to the 2014-08-20 version...
```

Alternatively, you can also create different Stripe clients.

#### Stripe Connect

If you are using Stripe Connect, you will want to make API calls on behalf of your client. You have nothing special
to do: just use the `setApiKey` method with the customer's access token:

```php
$client->setApiKey('my-customers-token');
// All API calls will be made on behalf of this customer now!
```

Please note that if you want to use again your own secret key, you will need to make a new call to this method.

#### Expand

All Stripe API requests have support to expand some nested objects inside responses. Therefore, ZfrStripe supports this
through the `expand` parameter, which must be an array:

```php
$details = $client->getCharges([
    'expand' => ['customer']
]);
```

#### Iterators

You may want to retrieve a list of customers, events or charges. Instead of manually doing all the requests yourself,
you can use iterators. ZfrStripe provides iterators for all iterable resources:

```php
$iterator = $client->getCustomersIterator();

foreach ($iterator as $user) {
    // Do something
}
```

By default, ZfrStripe retrieves 100 elements per API call (which is the maximum allowed by Stripe API). You may want
to lower this limit by using the `setPageSize` method. You can also set a upper bound of how many results you want
to retrieve by using the `setLimit` method.

Finally, you can still use API parameters when using an iterator. For instance, this will retrieve all the events
that have the event `customer.subscription.updated`, doing a new API call each 50 elements (this means that only
up to 50 elements are stored in memory), setting a limit of maximum 500 elements to retrieve:

```php
$iterator = $client->getEventsIterator(array('type' => 'customer.subscription.updated'));
$iterator->setPageSize(50);
$iterator->setLimit(500);

foreach ($iterator as $event) {
    // Do something
}
```

You can also take advantage of Stripe [cursor-based pagination](https://stripe.com/docs/api#pagination):

```php
$iterator = $client->getInvoicesIterator(['starting_after' => 'in_abcdef');

foreach ($iterator as $invoices) {
    // Do something
}
```

ZfrStripe takes care of fetching the last item in the batch, extracting the id, and continuing doing requests
until no more data is available!

#### Idempotency Key

Recently, Stripe has added support for idempotency key. This is added to every POST requests, and prevent an
operation for being done twice, as long as the key is the same. You are responsible to generate your own
idempotency key:

```php
$key = 'ABC';

$stripeClient->createSubscription([
    'id'              => 'cus_abc',
    'plan'            => 'planA',
    'idempotency_key' => $key
]);

$stripeClient->createSubscription([
    'id'              => 'cus_abc',
    'plan'            => 'planA',
    'idempotency_key' => $key
]);

// Only one subscription is created
```

#### Undocumented features

While playing with Stripe API, I realized that the API has a few filtering capabilities that were not documented,
but still accessible. Therefore, I've added some of them to this library but PLEASE be very defensive when you
use them, as Stripe may remove them (or maybe officially document them in the future). Here is a list of the
implemented hidden features:

##### Filter events

You can filter events by a object id using the `object_id` parameter. It will depend of the context. For instance
if you are specifically retrieving subscription events, the `object_id` will allow to filter by subscription:

```php
$events = $client->getEvents(['object_id' => 'cus_abc', 'type' => 'customer.*']);
```

##### Retrieve deleted customers

You can retrieve deleted customers by using the `deleted` boolean:

```php
$customers = $client->getCustomers(['deleted' => true]);
```

##### Retrieve non-paid charges

You can retrieve non paid charges by using the `paid` boolean:

```php
$notPaidCharges = $client->getCharges(['paid' => false]);
```

##### Retrieve proration items

You can retrieve proration invoice items only by using the `proration` boolean:

```php
$invoiceItems = $client->getInvoiceItems(['proration' => true]);
```

### Exceptions

ZfrStripe tries its best to throw useful exceptions. Two kinds of exceptions can occur:

* Guzzle exceptions: by default, Guzzle automatically validate parameters according to rules defined in the
 service description **before** sending the actual request. If you encounter those exceptions, you likely have broken
 code.
* ZfrStripe exceptions: those exceptions are thrown if an error occurred on Stripe side. Each exception implement
`ZfrStripe\Exception\ExceptionInterface`.

Here are all the exceptions:

* `ZfrStripe\Exception\UnauthorizedException`: your API key is likely to be wrong...
* `ZfrStripe\Exception\ApiRateLimitException`: occurs when you have reached the Stripe API limit.
* `ZfrStripe\Exception\BadRequestException`: occurs for 400 error code.
* `ZfrStripe\Exception\CardErrorException`: occurs for any card errors. According to Stripe, this error happen when
parameters were valid but the request failed (for instance if a card CVC is invalid).
* `ZfrStripe\Exception\NotFoundException`: is thrown whenever client receives a 404 exception.
* `ZfrStripe\Exception\RequestFailedException`: occurs for generic 402 error.
* `ZfrStripe\Exception\ServerErrorException`: any errors where Stripe is likely to be doing something wrong...

Usage:

```php
use ZfrStripe\Exception\TransactionErrorException;

try {
    $client->createCard([
        'card' => [
            'number'    => '4242424242424242',
            'exp_month' => '01',
            'exp_year'  => '2016'
        ]
    ]);
} catch (CardErrorException $exception) {
    // Seems we couldn't create the card, maybe because it's invalid
    $why = $exception->getMessage();

    // Let's also get the response to have more info:
    $response = $exception->getResponse();
} catch (\Exception $exception) {
    // Catch any other exception...
}
```

### Complete reference

Here is a complete list of all methods:

CHARGE RELATED METHODS:

* array captureCharge(array $args = array())
* array createCharge(array $args = array())
* array getCharge(array $args = array())
* array getCharges(array $args = array())
* array refundCharge(array $args = array())
* array updateCharge(array $args = array())

CUSTOMER RELATED METHODS:

* array createCustomer(array $args = array())
* array deleteCustomer(array $args = array())
* array getCustomer(array $args = array())
* array getCustomers(array $args = array())
* array updateCustomer(array $args = array())

CARD RELATED METHODS:

* array createCard(array $args = array())
* array deleteCard(array $args = array())
* array getCard(array $args = array())
* array getCards(array $args = array())
* array updateCard(array $args = array())

RECIPIENT CARD RELATED METHODS:

* array createRecipientCard(array $args = array())
* array deleteRecipientCard(array $args = array())
* array getRecipientCard(array $args = array())
* array getRecipientCards(array $args = array())
* array updateRecipientCard(array $args = array())

SUBSCRIPTION RELATED METHODS:

* array cancelSubscription(array $args = array())
* array createSubscription(array $args = array())
* array getSubscription(array $args = array())
* array getSubscriptions(array $args = array())
* array updateSubscription(array $args = array())

PLAN RELATED METHODS:

* array createPlan(array $args = array())
* array deletePlan(array $args = array())
* array getPlan(array $args = array())
* array getPlans(array $args = array())
* array updatePlan(array $args = array())

COUPON RELATED METHODS:

* array createCoupon(array $args = array())
* array deleteCoupon(array $args = array())
* array getCoupon(array $args = array())
* array getCoupons(array $args = array())
* array updateCoupon(array $args = array())

DISCOUNT RELATED METHODS:

* array deleteCustomerDiscount(array $args = array())
* array deleteSubscriptionDiscount(array $args = array())

INVOICE RELATED METHODS:

* array createInvoice(array $args = array())
* array getInvoice(array $args = array())
* array getInvoiceLineItems(array $args = array())
* array getInvoices(array $args = array())
* array getUpcomingInvoice(array $args = array())
* array getUpcomingInvoiceLineItems(array $args = array())
* array payInvoice(array $args = array())
* array updateInvoice(array $args = array())

INVOICE ITEM RELATED METHODS:

* array createInvoiceItem(array $args = array())
* array deleteInvoiceItem(array $args = array())
* array getInvoiceItem(array $args = array())
* array getInvoiceItems(array $args = array())
* array updateInvoiceItem(array $args = array())

DISPUTE RELATED METHODS:

* array getDispute(array $args = array()
* array getDisputes(array $args = array())
* array closeDispute(array $args = array())
* array updateDispute(array $args = array())

TRANSFER RELATED METHODS:

* array cancelTransfer(array $args = array())
* array createTransfer(array $args = array())
* array getTransfer(array $args = array())
* array getTransfers(array $args = array())
* array updateTransfer(array $args = array())

RECIPIENT RELATED METHODS:

* array createRecipient(array $args = array())
* array deleteRecipient(array $args = array())
* array getRecipient(array $args = array())
* array getRecipients(array $args = array())
* array updateRecipient(array $args = array())

REFUND RELATED METHODS:

* array getRefund(array $args = array())
* array getRefunds(array $args = array())
* array updateRefund(array $args = array())

APPLICATION FEE RELATED METHODS:

* array getApplicationFee(array $args = array())
* array getApplicationFees(array $args = array())
* array refundApplicationFee(array $args = array())

BALANCE RELATED METHODS:

* array getAccountBalance(array $args = array())
* array getBalanceTransaction(array $args = array())
* array getBalanceTransactions(array $args = array())

TOKEN RELATED METHODS:

* array createCardToken(array $args = array())
* array createBankAccountToken(array $args = array())
* array getToken(array $args = array())

EVENT RELATED METHODS:

* array getEvent(array $args = array())
* array getEvents(array $args = array())

ACCOUNT RELATED METHODS:

* array getAccount(array $args = array())


CHANGELOG.md
# 3.5.2

* Allow to add metadata when creating cards

# 3.5.1

* Allow to set metadata for cards when updating

# 3.5.0

* [BC] Set default Stripe API to "2015-10-16"
* Add the "default_source" parameter when updating a Stripe customer (all descriptors).

# 3.4.0

* [BC] Set default Stripe API to "2015-10-01"

# 3.3.0

* [BC] Set default Stripe API to "2015-09-08"

# 3.2.0

* [BC] Set default Stripe API to "2015-09-03"

# 3.1.0

* [BC] Set default Stripe API to "2015-08-19". When using this new version, refunds are now get without specifying
a charge ID.

# 3.0.0

* [BC] For all descriptors, `closeDispute` and `updateDispute` now uses the new endpoints that do not require a charge ID but instead a dispute
ID. In previous versions:

```php
$stripeClient->closeDispute(['charge' => 'ch_abc']);
```

In new version:

```php
$stripeClient->closeDispute(['id' => 'dp_abc']);
```

Also, new endpoints have been added that allow to retrieve a specific dispute by its id, as well as retrieving all disputes.

* [BC] Descriptor cleanup: older descriptors have been removed. ZfrStripe now supports Stripe API version from 2015-02-18. If you are using an older
version of the API, you should stay on 2.x branch.

# 2.20.0

* [BC] Set default Stripe API to "2015-07-28"

# 2.19.0

* [BC] Set default Stripe API to "2015-07-13"

# 2.18.0

* [BC] Set default Stripe API to "2015-07-07"

# 2.17.0

* [BC] Set default Stripe API to "2015-06-15"

# 2.16.1

* Add the `destination` parameter when creating a charge (in all descriptors).

# 2.16.0

* [BC] Set default Stripe API to "2015-04-07" version.

# 2.15.0

* [BC] Set default Stripe API to "2015-03-24" version. This adds the property `discountable` for `createInvoiceItem`
and `updateInvoiceItem` for all descriptors.

# 2.14.1

* Fix missing import

# 2.14.0

* [BC] Set default Stripe API to "2015-02-18" version (note: this version brings several important breaking changes
in how you use Stripe API).
* Fix a bug when changing the API version at runtime did not properly updated the descriptor.

# 2.13.0

* [BC] Set default Stripe API to "2015-02-16" version

# 2.12.1

* Set a simpler dependency for Guzzle (hopefully solving some dependencies problems)

# 2.12.0

* [BC] Set default Stripe API to "2015-02-10" version

# 2.11.1

* Validation for `tax_percent` is less restrictive and now accepts both a string or float

# 2.11.0

* [BC] Set default Stripe API to "2015-01-26" version

# 2.10.2

* ZfrStripe adds support for the new Idempotency feature ([see here](https://stripe.com/docs/api#errors)). It is added
to every POST request, and work by using the new `idempotency_key` param.

# 2.10.1

* Fix a bug where the parameter `plan` was required for `updateSubscription` (it is actually optional).

# 2.10.0

* [BC] Set default Stripe API to "2015-01-11" version

# 2.9.0

* Fixed a long-lasting bug for `updateDispute` method. Also, a new descriptor has been added, as a previous BC
was not taken into account in ZfrStripe. Therefore, current descriptor is Stripe-v1.2.

# 2.8.1

* Add the new `tax_percent` property to subscriptions.

# 2.8.0

* [BC] Set default Stripe API to "2014-12-22" version

# 2.7.0

* [BC] Set default Stripe API to "2014-12-17" version. This version replaced the `statement_description` to
`statement_descriptor` for charges, plans, invoices and tranfers. Therefore, a new descriptor has been released for
versions starting at "2014-12-17".

# 2.6.0

* [BC] Set default Stripe API to "2014-12-08" version

# 2.5.0

* [BC] Set default Stripe API to "2014-11-20" version

# 2.4.0

* [BC] Set default Stripe API to "2014-11-05" version

# 2.3.6

* Add support for the new `reason` property when creating a refund.

# 2.3.5

* Fix IDE auto-completion hint for `getUpcomingInvoiceLineItems` method

# 2.3.4

* Add missing `getUpcomingInvoiceLineItems` endpoint. This can be used to retrieve all the invoice line items from
a given upcoming invoice.

# 2.3.3

* `forgiven` boolean was not properly encoded when updating an invoice.

# 2.3.2

* Add support for filtering deleted customers or not through the `deleted` parameter. Please note
that this feature **IS NOT** documented and officially supported by Stripe, and may be removed without caution.

# 2.3.1

* Add support for the new `shipping` array when creating charges.
* Add support for filtering charges if they are paid or not through the `paid` parameter. Please note
that this feature **IS NOT** documented and officially supported by Stripe, and may be removed without caution.

# 2.3.0

* [BC] Set default Stripe API to "2014-10-07" version
* ZfrStripe now uses PSR-4 for autoloading. This has flattened the structure of the libray, it should not break any of your code :).

# 2.2.4

* You can now use the shortcut `now` when setting or updating `trial_end` property.
* Add support for the new `billing_cycle_anchor` property when creating and updating a subscription.

# 2.2.3

* Fix a bug when filtering invoice items by date.

# 2.2.2

* Add support for filtering invoice items if they are proration or not through the `proration` parameter. Please note
that this feature **IS NOT** documented and officially supported by Stripe, and may be removed without caution.

# 2.2.1

* Add support for `object_id` when retrieving events. This allows to filter events by customer. Please note that
this feature **IS NOT** documented and officially supported by Stripe, but as it powers Stripe dashboard, it should
not be removed.

# 2.2.0

* The `getInvoiceLineItems` was not working correctly when pagination.
* [BC] For consistency with other resources, the `id` parameter for retrieving invoice line items has been rename
`invoice`.

# 2.1.0

* [BC] Set default Stripe API to "2014-09-08" version.

# 2.0.4

* Fix docblock to allow auto-completion on `getInvoiceLineItemsIterator`

# 2.0.3

* Balance transactions history for fee refund can be retrieved using the "fee_refund" type (Stripe documentation was
wrong)

# 2.0.2


src/Client/StripeClient.php
<?php
/*
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software consists of voluntary contributions made by many individuals
 * and is licensed under the MIT license.
 */

namespace ZfrStripe\Client;

use Guzzle\Common\Event;
use Guzzle\Plugin\ErrorResponse\ErrorResponsePlugin;
use Guzzle\Service\Client;
use Guzzle\Service\Command\Factory\CompositeFactory;
use Guzzle\Service\Description\ServiceDescription;
use Guzzle\Service\Resource\ResourceIterator;
use ZfrStripe\Client\Iterator\StripeCommandsCursorIterator;
use ZfrStripe\Exception\UnsupportedStripeVersionException;
use ZfrStripe\Http\QueryAggregator\StripeQueryAggregator;

/**
 * @author  Michaël Gallego <mic.gallego@gmail.com>
 * @licence MIT
 *
 * CHARGE RELATED METHODS:
 *
 * @method array captureCharge(array $args = array()) {@command Stripe CaptureCharge}
 * @method array createCharge(array $args = array()) {@command Stripe CreateCharge}
 * @method array getCharge(array $args = array()) {@command Stripe GetCharge}
 * @method array getCharges(array $args = array()) {@command Stripe GetCharges}
 * @method array refundCharge(array $args = array()) {@command Stripe RefundCharge}
 * @method array updateCharge(array $args = array()) {@command Stripe UpdateCharge}
 *
 * CUSTOMER RELATED METHODS:
 *
 * @method array createCustomer(array $args = array()) {@command Stripe CreateCustomer}
 * @method array deleteCustomer(array $args = array()) {@command Stripe DeleteCustomer}
 * @method array getCustomer(array $args = array()) {@command Stripe GetCustomer}
 * @method array getCustomers(array $args = array()) {@command Stripe GetCustomers}
 * @method array updateCustomer(array $args = array()) {@command Stripe UpdateCustomer}
 *
 * CARD RELATED METHODS:
 *
 * @method array createCard(array $args = array()) {@command Stripe CreateCard}
 * @method array deleteCard(array $args = array()) {@command Stripe DeleteCard}
 * @method array getCard(array $args = array()) {@command Stripe GetCard}
 * @method array getCards(array $args = array()) {@command Stripe GetCards}
 * @method array updateCard(array $args = array()) {@command Stripe UpdateCard}
 *
 * RECIPIENT CARD RELATED METHODS:
 *
 * @method array createRecipientCard(array $args = array()) {@command Stripe CreateRecipientCard}
 * @method array deleteRecipientCard(array $args = array()) {@command Stripe DeleteRecipientCard}
 * @method array getRecipientCard(array $args = array()) {@command Stripe GetRecipientCard}
 * @method array getRecipientCards(array $args = array()) {@command Stripe GetRecipientCards}
 * @method array updateRecipientCard(array $args = array()) {@command Stripe UpdateRecipientCard}
 *
 * SUBSCRIPTION RELATED METHODS:
 *
 * @method array cancelSubscription(array $args = array()) {@command Stripe CancelSubscription}
 * @method array createSubscription(array $args = array()) {@command Stripe CreateSubscription}
 * @method array getSubscription(array $args = array()) {@command Stripe GetSubscription}
 * @method array getSubscriptions(array $args = array()) {@command Stripe GetSubscriptions}
 * @method array updateSubscription(array $args = array()) {@command Stripe UpdateSubscription}
 *
 * PLAN RELATED METHODS:
 *
 * @method array createPlan(array $args = array()) {@command Stripe CreatePlan}
 * @method array deletePlan(array $args = array()) {@command Stripe DeletePlan}
 * @method array getPlan(array $args = array()) {@command Stripe GetPlan}
 * @method array getPlans(array $args = array()) {@command Stripe GetPlans}
 * @method array updatePlan(array $args = array()) {@command Stripe UpdatePlan}
 *
 * COUPON RELATED METHODS:
 *
 * @method array createCoupon(array $args = array()) {@command Stripe CreateCoupon}
 * @method array deleteCoupon(array $args = array()) {@command Stripe DeleteCoupon}
 * @method array getCoupon(array $args = array()) {@command Stripe GetCoupon}
 * @method array getCoupons(array $args = array()) {@command Stripe GetCoupons}
 * @method array updateCoupon(array $args = array()) {@command Stripe UpdateCoupon}
 *
 * DISCOUNT RELATED METHODS:
 *
 * @method array deleteCustomerDiscount(array $args = array()) {@command Stripe DeleteCustomerDiscount}
 * @method array deleteSubscriptionDiscount(array $args = array()) {@command Stripe DeleteSubscriptionDiscount}
 *
 * INVOICE RELATED METHODS:
 *
 * @method array createInvoice(array $args = array()) {@command Stripe CreateInvoice}
 * @method array getInvoice(array $args = array()) {@command Stripe GetInvoice}
 * @method array getInvoiceLineItems(array $args = array()) {@command Stripe GetInvoiceLineItems}
 * @method array getInvoices(array $args = array()) {@command Stripe GetInvoices}
 * @method array getUpcomingInvoice(array $args = array()) {@command Stripe GetUpcomingInvoice}
 * @method array getUpcomingInvoiceLineItems(array $args = array()) {@command Stripe GetUpcomingInvoiceLineItems}
 * @method array payInvoice(array $args = array()) {@command Stripe PayInvoice}
 * @method array updateInvoice(array $args = array()) {@command Stripe UpdateInvoice}
 *
 * INVOICE ITEM RELATED METHODS:
 *
 * @method array createInvoiceItem(array $args = array()) {@command Stripe CreateInvoiceItem}
 * @method array deleteInvoiceItem(array $args = array()) {@command Stripe DeleteInvoiceItem}
 * @method array getInvoiceItem(array $args = array()) {@command Stripe GetInvoiceItem}
 * @method array getInvoiceItems(array $args = array()) {@command Stripe GetInvoiceItems}
 * @method array updateInvoiceItem(array $args = array()) {@command Stripe UpdateInvoiceItem}
 *
 * DISPUTE RELATED METHODS:
 *
 * @method array getDispute(array $args = array()) {@command Stripe GetDispute}
 * @method array getDisputes(array $args = array()) {@command Stripe GetDisputes}
 * @method array closeDispute(array $args = array()) {@command Stripe CloseDispute}
 * @method array updateDispute(array $args = array()) {@command Stripe UpdateDispute}
 *
 * TRANSFER RELATED METHODS:
 *
 * @method array cancelTransfer(array $args = array()) {@command Stripe CancelTransfer}
 * @method array createTransfer(array $args = array()) {@command Stripe CreateTransfer}
 * @method array getTransfer(array $args = array()) {@command Stripe GetTransfer}
 * @method array getTransfers(array $args = array()) {@command Stripe GetTransfers}
 * @method array updateTransfer(array $args = array()) {@command Stripe UpdateTransfer}
 *
 * RECIPIENT RELATED METHODS:
 *
 * @method array createRecipient(array $args = array()) {@command Stripe CreateRecipient}
 * @method array deleteRecipient(array $args = array()) {@command Stripe DeleteRecipient}
 * @method array getRecipient(array $args = array()) {@command Stripe GetRecipient}
 * @method array getRecipients(array $args = array()) {@command Stripe GetRecipients}
 * @method array updateRecipient(array $args = array()) {@command Stripe UpdateRecipient}
 *
 * REFUND RELATED METHODS:
 *
 * @method array getRefund(array $args = array()) {@command Stripe GetRefund}
 * @method array getRefunds(array $args = array()) {@command Stripe GetRefunds}
 * @method array updateRefund(array $args = array()) {@command Stripe UpdateRefunds}
 *
 * APPLICATION FEE RELATED METHODS:
 *
 * @method array getApplicationFee(array $args = array()) {@command Stripe GetApplicationFee}
 * @method array getApplicationFees(array $args = array()) {@command Stripe GetApplicationFees}
 * @method array refundApplicationFee(array $args = array()) {@command Stripe RefundApplicationFee}
 *
 * APPLICATION FEE REFUND RELATED METHODS:
 *
 * @method array getApplicationFeeRefund(array $args = array()) {@command Stripe GetApplicationFeeRefund}
 * @method array getApplicationFeeRefunds(array $args = array()) {@command Stripe GetApplicationFeeRefunds}
 * @method array updateApplicationFeeRefund(array $args = array()) {@command Stripe UpdateApplicationFeeRefund}
 *
 * BALANCE RELATED METHODS:
 *
 * @method array getAccountBalance(array $args = array()) {@command Stripe GetAccountBalance}
 * @method array getBalanceTransaction(array $args = array()) {@command Stripe GetBalanceTransaction}
 * @method array getBalanceTransactions(array $args = array()) {@command Stripe GetBalanceTransactions}
 *
 * TOKEN RELATED METHODS:
 *
 * @method array createCardToken(array $args = array()) {@command Stripe CreateCardToken}
 * @method array createBankAccountToken(array $args = array()) {@command Stripe CreateBankAccountToken}
 * @method array getToken(array $args = array()) {@command Stripe GetToken}
 *
 * EVENT RELATED METHODS:
 *
 * @method array getEvent(array $args = array()) {@command Stripe GetEvent}
 * @method array getEvents(array $args = array()) {@command Stripe GetEvents}
 *
 * ACCOUNT RELATED METHODS:
 *
 * @method array getAccount(array $args = array()) {@command Stripe GetAccount}
 *
 * ITERATOR METHODS:
 *
 * @method ResourceIterator getCustomersIterator()
 * @method ResourceIterator getChargesIterator()
 * @method ResourceIterator getCardsIterator()
 * @method ResourceIterator getRecipientCardsIterator()
 * @method ResourceIterator getSubscriptionsIterator()
 * @method ResourceIterator getPlansIterator()
 * @method ResourceIterator getCouponsIterator()
 * @method ResourceIterator getInvoicesIterator()
 * @method ResourceIterator getInvoiceLineItemsIterator()
 * @method ResourceIterator GetUpcomingInvoiceLineItemsIterator()
 * @method ResourceIterator getInvoiceItemsIterator()
 * @method ResourceIterator getDisputesIterator()
 * @method ResourceIterator getTransfersIterator()
 * @method ResourceIterator getRecipientsIterator()
 * @method ResourceIterator getRefundsIterator()
 * @method ResourceIterator getApplicationFeesIterator()
 * @method ResourceIterator getApplicationFeeRefundsIterator()
 * @method ResourceIterator getBalanceTransactionsIterator()
 * @method ResourceIterator getEventsIterator()
 */
class StripeClient extends Client
{
    /**
     * Stripe API version
     */
    const LATEST_API_VERSION = '2015-10-16';

    /**
     * @var array
     */
    protected $availableVersions = [
        '2015-02-18', '2015-03-24', '2015-04-07', '2015-06-15', '2015-07-07', '2015-07-13', '2015-07-28',
        '2015-08-07', '2015-08-19', '2015-09-03', '2015-09-08', '2015-09-23', '2015-10-01', '2015-10-12',
        '2015-10-16'
    ];

    /**
     * @var string
     */
    protected $apiKey;

    /**
     * @var string
     */
    protected $version;

    /**
     * Constructor
     *
     * @param string $apiKey
     * @param string $version
     */
    public function __construct($apiKey, $version = self::LATEST_API_VERSION)
    {
        parent::__construct();

        $this->setApiKey($apiKey);
        $this->setApiVersion($version);

        $this->setUserAgent('zfr-stripe-php', true);

        // Add an event to set the Authorization param before sending any request
        $dispatcher = $this->getEventDispatcher();

        $dispatcher->addSubscriber(new ErrorResponsePlugin());
        $dispatcher->addListener('command.after_prepare', [$this, 'afterPrepare']);
        $dispatcher->addListener('command.before_send', [$this, 'authorizeRequest']);
    }

    /**
     * Override the API key
     *
     * @param  string $apiKey
     * @return void
     */
    public function setApiKey($apiKey)
    {
        $this->apiKey = (string) $apiKey;
    }

    /**
     * Get the current API key
     *
     * @return string
     */
    public function getApiKey()
    {
        return $this->apiKey;
    }

    /**
     * {@inheritdoc}
     */
    public function __call($method, $args = [])
    {
        if (substr($method, -8) === 'Iterator') {
            // Allow magic method calls for iterators (e.g. $client-><CommandName>Iterator($params))
            $commandOptions  = isset($args[0]) ? $args[0] : [];
            $iteratorOptions = isset($args[1]) ? $args[1] : [];
            $command         = $this->getCommand(substr($method, 0, -8), $commandOptions);

            return new StripeCommandsCursorIterator($command, $iteratorOptions);
        }

        return parent::__call(ucfirst($method), $args);
    }

    /**
     * Set the Stripe API version
     *
     * This method forces to add a Stripe-Version header, so that you can use the wanted version no matter
     * what the setting is on Stripe
     *
     * @param  string $version
     * @return void
     */
    public function setApiVersion($version)
    {
        if (!in_array($version, $this->availableVersions, true)) {
            throw new UnsupportedStripeVersionException(sprintf(
                'The given Stripe version ("%s") is not supported by ZfrStripe. Value must be one of the following: "%s"',
                $version,
                implode(', ', $this->availableVersions)
            ));
        }

        $this->version = (string) $version;
        $this->setDefaultOption('headers', ['Stripe-Version' => $this->version]);

        if ($this->version < '2015-08-19') {
            $descriptor = __DIR__ . '/ServiceDescription/Stripe-v1.0.php';
        } else {
            $descriptor = __DIR__ . '/ServiceDescription/Stripe-v1.1.php';
        }

        // Note that we need to recreate a whole new CompositeFactory whenever the version is changed
        $this->setCommandFactory(new CompositeFactory());
        $this->setDescription(ServiceDescription::factory($descriptor));
    }

    /**
     * Get current Stripe API version
     *
     * @return string
     */
    public function getApiVersion()
    {
        return $this->version;
    }

    /**
     * Modify the query aggregator
     *
     * @internal
     * @param  Event $event
     * @return void
     */
    public function afterPrepare(Event $event)
    {
        /* @var \Guzzle\Service\Command\CommandInterface $command */
        $command = $event['command'];
        $request = $command->getRequest();

        $request->getQuery()->setAggregator(new StripeQueryAggregator());
    }

    /**
     * Authorize the request
     *
     * @internal
     * @param  Event $event
     * @return void
     */
    public function authorizeRequest(Event $event)
    {
        /* @var \Guzzle\Service\Command\CommandInterface $command */
        $command = $event['command'];

        $request = $command->getRequest();
        $request->setAuth($this->apiKey);
    }
}
